src/Infrastructure/LeanCode.SendGrid/SendGridException.cs

```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Net;
+using System.Text.Json;
+using System.Threading.Tasks;
+using SendGrid;
+
+namespace LeanCode.SendGrid
+{
+    public class SendGridException : Exception
+    {
+        private static readonly JsonSerializerOptions SerializerOptions = new JsonSerializerOptions
+        {
+            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+        };
+
+        public HttpStatusCode ResponseStatusCode { get; }
+        public ImmutableArray<string> ErrorMessages { get; } = ImmutableArray<string>.Empty;
+
+        private SendGridException(
+            HttpStatusCode responseStatusCode,
+            IEnumerable<SendGridError>? errors)
+            : base($"SendGrid request failed with status code {responseStatusCode}.")
+        {
+            ResponseStatusCode = responseStatusCode;
+
+            if (errors is object)
+            {
+                ErrorMessages = errors
+                    .Select(e => e.Message!)
+                    .Where(m => m is object)
+                    .ToImmutableArray();
+            }
+        }
+
+        public static async Task ThrowOnFailureAsync(Response response)
```

Comment:
Is this a common pattern? I haven't really seen having these kind of helper methods in `Exception` classes (and I don't think this is the place for these methods).

Reply:
Maybe not, but where would you move such logic anyway?

Reply:
In the client - it is responsible for doing requests and handling results.